packages/clay-core/src/tree-view/useItem.tsx

```diff
@@ -0,0 +1,88 @@
+/**
+ * SPDX-FileCopyrightText: © 2021 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import React, {useContext, useEffect, useRef} from 'react';
+import {useDrag, useDrop} from 'react-dnd';
+
+import {useTreeViewContext} from './context';
+
+type Value = {
+	[propName: string]: any;
+	indexes: Array<number>;
+	key: React.Key;
+};
+
+type Props = {
+	children: React.ReactNode;
+	value: Omit<Value, 'indexes'> & {index: number};
+};
+
+const ItemContext = React.createContext<Value>({} as Value);
+
+function getKey(key: React.Key) {
+	return `${key}`.replace('.$', '');
+}
+
+export function ItemContextProvider({children, value}: Props) {
+	const {items, open, reorder, selection} = useTreeViewContext();
+	const {indexes: parentIndexes = [], key: parentKey} = useItem();
+
+	const keyRef = useRef(getKey(value.key));
+
+	const childRef = useRef(null);
+
+	useEffect(() => selection.mount(keyRef.current, parentKey), [
+		selection.mount,
+		keyRef,
+		parentKey,
+	]);
+
+	const item: Value = {
+		...value,
+		indexes: [...parentIndexes, value.index],
+		key: keyRef.current,
+	};
+
+	const [, drag] = useDrag({
```

Comment:
Yes good point, I'll have a look into it.

Reply:
Hey @matuzalemsteles about the preview (sorry I'm not able to paste the image from the Google Document here), what's your idea? I was thinking I'd create an image representing the "node being dragged" from a `canvas` element and using `canvas.toBlob` or similar and pass it to the preview component. Does that make sense to you?

Reply:
I think we can just create an HTML markup 😅, for example we can render in DnD preview an empty image and use `useDragLayer` to get the position and render the markup.

We added this to `useItem` because of the `useDrag` which returns the `preview`.

```js
useEffect(() => {
  preview(getEmptyImage(), {captureDraggingState: true});
}, [preview]);
```

And we created a `DragLayer` component that would be the preview, something like this:

```js
const layerStyles = {
  cursor: 'grabbing',
  height: '100%',
  left: 0,
  pointerEvents: 'none',
  position: 'fixed',
  top: 0,
  width: '100%',
  zIndex: 150,
};

function getItemStyles(initialOffset, currentOffset) {
  if (!initialOffset || !currentOffset) {
    return {
      display: 'none',
    };
  }

  const {x, y} = currentOffset;
  const transform = `translate(${x}px, ${y}px)`;

  return {
    WebkitTransform: transform,
    transform,
  };
}

const DragLayer = () => {
  const {currentOffset, initialOffset, isDragging, item} = useDragLayer(
    (monitor) => ({
      currentOffset: monitor.getSourceClientOffset(),
      initialOffset: monitor.getInitialSourceClientOffset(),
      isDragging: monitor.isDragging(),
      item: monitor.getItem(),
    })
  );

  if (!isDragging) {
    return null;
  }

  return (
    <div style={layerStyles}>
      <div
        className="some-class-to-style"
        style={getItemStyles(initialOffset, currentOffset, isDragging)}
      >
        {item.name}
      </div>
    </div>
  );
};
```

We can call this component in `TreeView`. This is an example I made in DXP a while ago with some modifications.

Reply:
Thanks @matuzalemsteles I'll test that.

Reply:
Added a "custom" drag layer based on this in https://github.com/liferay/clay/pull/4255/commits/2c1b354a9bed512e549a8392bbe0aadab37d3a12 we'll just need to add that `some-class-to-style` to make it look a bit nicer.